designs/validations.md

```diff
@@ -0,0 +1,288 @@
+
+```go
+// Sequentially composes a set of validations into one which will run them sequentially and in order.
+func Sequentially(validations ...Validation) Validation
+```
+
+It can be tempting to allow to compose concurrent validations inside a sequential collection. Although appealing, this could significantly increase the complexity of the runner. Since we can't find a place where this is currently needed, the framework will only run concurrently the first level validations. If that ever becomes a requirement, we will default to wrap another runner inside a handler. Given all returned aggregated errors are flatten, this should work without changes to the framework.
+
+#### Logging
+Some of the current attempts at a validation runner/composer log the success of validations. This requires providing a name for the validation. In addition, some of the provider validations log both success and failure of validations (using ✅  and ❌).
+
+Trying to handle this at the framework level has several drawbacks:
+* It requires naming validations, which increases the framework surface area, increases the entry level to register a validation and composability more difficult.
+* It limits logging granularity when a validation function is composed of multiple sequential validations.
+
+For these reasons, the framework won't provide any logging functionality.
+	* For debug logging (we can generalize this to anything above verbosity `0`), validations should use their own logger. This should be injected at construction time as any other dependency. This still gives control to the program to decide which logs are presented to the user and when, by passing different loggers.
+	* For information more tailored to feedback for the end user (like progress in slow validations), it will be left up to the framework callers to implement this. Based how logging is currently used to communicate this progress, the information is usually presented by "grouping" validations and not at the individual validation level. So the *usecases*, where validations are grouped and given semantic value, will be responsible to define these different groups and print logs in between them as necessary. This can be easily accomplished by composing multiple runners. Given a traditional logger might not be the best way to implement a presentation layer of this kind, this approach also allows to replace the logger with something else (in the future) without having to refactor any of the validation logic.
+
+#### Validation error meta information
+The existing `validations.ValidationResult` allows handlers to return a `Remediation` message. Given we want to just use the `error` interface to make it easy to register validations and that `Remediation` is empty for most `ValidationResult`'s , we chose to make this optional. Validations wanting to provide a possible remediation for a failure, can return an `error` that also implements the `Fixable` interface.
```

Comment:
```suggestion
The existing `validations.ValidationResult` allows handlers to return a `Remediation` message. Given we want to just use the `error` interface to make it easy to register validations and that `Remediation` is empty for most `ValidationResult`s , we chose to make this optional. Validations wanting to provide a possible remediation for a failure, can return an `error` that also implements the `Fixable` interface.
```

Reply:
Also I am confused by this, so where is `Fixable` returned for the validator runner to process remediation and print? Or is that the job of the framework callers?

Reply:
I think I see it now in the example below, but essentially we check for whether this is a Fixable interface or not with remediation?

Reply:
> I think I see it now in the example below, but essentially we check for whether this is a Fixable interface or not with remediation?

That's the gist.

Reply:
The runner doesn't process errors or print them or do anything with them except flatten them if they are aggregated. It's up to the caller to decide what to do with those errors.

And exactly, we offer a util method that inspects errors to see if they implement `Fixable` or not, and if they do, return the remediation message. That is supposed to be used by the caller, in case it's interested in that kind of info.